### PR TITLE
 refactor(core): avoid object creation in bind

### PR DESCRIPTION
**Pull request by kara**
Implementing optimization @mhevery and I talked about last week to avoid creating unnecessary objects (causes GC pressure). 

Notes: 

- Review only last commit. 

- To avoid unnecessary addition operations (which may be slower on mobile), the indices are saved once they are calculated. This adds to code size, but shouldn't have significant impact once minified.

### CODE REVIEW

**Comment by pkozlowski-opensource** [comment]
The change LGTM, but I would like to better understand this part:

>  to avoid creating unnecessary objects (causes GC pressure).

If I'm reading the code correctly, this is about creation mode, right? We are avoiding creating an object that would store old values (ex. `data[bindingIndex++] = {v0: v0, v1: v1};`). But at the same time we are creating more entries (elements) in the `data` array (a separate entry for each and every "old value"). Provided that I got the intention right, we are trading object creation for entries in array creation?

**Comment by mhevery** [comment]
So each object has an overhead 56 bytes plus any properties stored an it. So moving from object to Array saves 56 bytes per binding. (This is because we already payed the array overhead).

There is an additional benefit in that when reading location `x` in array we will cause a cache fetch which means that  `x+1` read will most likely result in cache hit. So we get benefit of locality. We also save the cost of dereferencing through an un-needed object.

File: packages/core/src/render3/instructions.ts [hunk 1]
```diff
@@ -1387,14 +1387,18 @@ export function bind2(prefix: string, v0: any, i0: string, v1: any, suffix: stri
...
   } else {
-    const parts: {v0: any, v1: any} = data[bindingIndex++];
-    if (v0 === NO_CHANGE) v0 = parts.v0;
-    if (v1 === NO_CHANGE) v1 = parts.v1;
-    if (different = (isDifferent(parts.v0, v0) || isDifferent(parts.v1, v1))) {
-      parts.v0 = v0;
-      parts.v1 = v1;
+    const part0Index = bindingIndex++;
+    const part1Index = bindingIndex++;
+    const part0 = data[part0Index];
```
Comment (mhevery) [inline packages/core/src/render3/instructions.ts]: I would inline the `++` as `data[bindingIndex++]`. This saves the creation of local war. (See next comment)

File: packages/core/src/render3/instructions.ts [hunk 2]
```diff
@@ -1387,14 +1387,18 @@ export function bind2(prefix: string, v0: any, i0: string, v1: any, suffix: stri
...
+    const part1 = data[part1Index];
+    if (v0 === NO_CHANGE) v0 = part0;
+    if (v1 === NO_CHANGE) v1 = part1;
+    if (different = (isDifferent(part0, v0) || isDifferent(part1, v1))) {
+      data[part0Index] = v0;
```
Comment (mhevery) [inline packages/core/src/render3/instructions.ts]: here just do `data[bingIndex - 2]` since stores are rare, this is fine

**Review by mhevery — CHANGES_REQUESTED** [review]

**Comment by kara** [comment]
@mhevery  Should be ready

**Comment by kara** [comment]
Closed in favor of https://github.com/angular/angular/pull/21155

**Pull request closed without merging**